plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,196 @@
+        } else {
+            fetchAllRules(attributeFilters, listener);
+        }
+    }
+
+    /**
+     * Fetch a single rule from system index using id
+     * @param id - The id of the rule to get.
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    private void fetchRuleById(String id, ActionListener<GetRuleResponse> listener) {
+        ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext().stashContext();
+        client.prepareGet(RULE_INDEX, id)
+            .execute(ActionListener.wrap(
+                getResponse -> {
+                    try (ThreadContext.StoredContext context = storedContext) {
+                        handleGetOneRuleResponse(id, getResponse, listener);
+                    }
+                },
+                e -> {
+                    try (ThreadContext.StoredContext context = storedContext) {
+                        logger.error("Failed to fetch rule with ID {}: {}", id, e.getMessage());
+                        listener.onFailure(e);
+                    }
+                }
+            ));
+    }
+
+    /**
+     * Process getResponse from index and send a GetRuleResponse
+     * @param id - The id of the rule to get
+     * @param getResponse - Response received from index
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    private void handleGetOneRuleResponse(String id, GetResponse getResponse, ActionListener<GetRuleResponse> listener) {
+        if (getResponse.isExists()) {
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                XContentParser parser = MediaTypeRegistry.JSON.xContent()
+                    .createParser(
+                        NamedXContentRegistry.EMPTY,
+                        DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                        getResponse.getSourceAsString()
+                    );
+                listener.onResponse(new GetRuleResponse(Map.of(id, Builder.fromXContent(parser).build()), RestStatus.OK));
+            } catch (IOException e) {
+                logger.error("Error parsing rule with ID {}: {}", id, e.getMessage());
+                listener.onFailure(e);
+            }
+        } else {
+            listener.onFailure(new ResourceNotFoundException("Rule with ID " + id + " not found."));
+        }
+    }
+
+    /**
+     * Fetch all rule from system index based on attributeFilters
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    private void fetchAllRules(Map<RuleAttribute, Set<String>> attributeFilters, ActionListener<GetRuleResponse> listener) {
+        ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext().stashContext();
+        client.prepareSearch(RULE_INDEX)
+            .setQuery(QueryBuilders.matchAllQuery())
```

Comment:
This query should contain all the filters and query clauses, that will help us remove majority of the logic in this class